akka-stream/src/main/scala/akka/stream/impl/SetupStage.scala

```diff
@@ -0,0 +1,148 @@
+    subOutlet.setHandler(delegateToInlet(() ⇒ pull(in), () ⇒ cancel(in)))
+
+    setHandler(in, delegateToSubOutlet(() ⇒ grab(in), subOutlet))
+    setHandler(out, delegateToSubInlet(subInlet))
+
+    override def preStart(): Unit = {
+      val flow = factory(ActorMaterializerHelper.downcast(materializer))(attributes)
+
+      val mat = Source
+        .fromGraph(subOutlet.source)
+        .viaMat(flow.withAttributes(attributes))(Keep.right)
+        .to(Sink.fromGraph(subInlet.sink))
+        .run()(subFusingMaterializer)
+      matPromise.success(mat)
+    }
+  }
+}
+
+/** Internal Api */
+@InternalApi private[stream] final class SetupSourceStage[T, M](factory: ActorMaterializer ⇒ Attributes ⇒ Source[T, M])
+  extends GraphStageWithMaterializedValue[SourceShape[T], Future[M]] {
+
+  private val out = Outlet[T]("SetupSourceStage.out")
+  override val shape = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    (createStageLogic(matPromise), matPromise.future)
+  }
+
+  private def createStageLogic(matPromise: Promise[M]) = new GraphStageLogic(shape) {
+    import SetupStage._
+
+    val subInlet = new SubSinkInlet[T]("SetupSourceStage")
+    subInlet.setHandler(delegateToOutlet(push(out, _: T), () ⇒ complete(out), fail(out, _), subInlet))
+    setHandler(out, delegateToSubInlet(subInlet))
+
+    override def preStart(): Unit = {
+      val source = factory(ActorMaterializerHelper.downcast(materializer))(attributes)
+
+      val mat = source
+        .withAttributes(attributes)
+        .to(Sink.fromGraph(subInlet.sink))
+        .run()(subFusingMaterializer)
+      matPromise.success(mat)
+    }
+  }
+}
+
+private object SetupStage {
+  def delegateToSubOutlet[T](grab: () ⇒ T, subOutlet: GraphStageLogic#SubSourceOutlet[T]) = new InHandler {
+    override def onPush(): Unit =
+      subOutlet.push(grab())
+    override def onUpstreamFinish(): Unit =
+      subOutlet.complete()
+    override def onUpstreamFailure(ex: Throwable): Unit =
+      subOutlet.fail(ex)
+  }
+
+  def delegateToOutlet[T](
+    push:     T ⇒ Unit,
+    complete: () ⇒ Unit,
+    fail:     Throwable ⇒ Unit,
+    subInlet: GraphStageLogic#SubSinkInlet[T]) = new InHandler {
+    override def onPush(): Unit =
+      push(subInlet.grab())
+    override def onUpstreamFinish(): Unit =
+      complete()
+    override def onUpstreamFailure(ex: Throwable): Unit =
+      fail(ex)
+  }
+
+  def delegateToSubInlet[T](subInlet: GraphStageLogic#SubSinkInlet[T]) = new OutHandler {
+    override def onPull(): Unit =
+      subInlet.pull()
+    override def onDownstreamFinish(): Unit =
+      subInlet.cancel()
+  }
+
+  def delegateToInlet(pull: () ⇒ Unit, cancel: () ⇒ Unit) = new OutHandler {
+    override def onPull(): Unit =
+      pull()
+    override def onDownstreamFinish(): Unit =
+      cancel()
+  }
+}
```

Comment:
Perhaps not in the scope of this but I wonder if we could make this something closer to the materializer and avoid the submaterialization.